clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -660,6 +670,85 @@ public void subscribe(TopicPartition... partitions) {
     }
 
     /**
+     * Subscribes to topics matching specified pattern and uses the consumer's group
+     * management functionality. The pattern matching will be done periodically against topics
+     * existing at the time of check.
+     * <p>
+     * As part of group management, the consumer will keep track of the list of consumers that
+     * belong to a particular group and will trigger a rebalance operation if one of the
+     * following events trigger -
+     * <ul>
+     * <li>Number of partitions change for any of the subscribed list of topics
+     * <li>Topic is created or deleted
+     * <li>An existing member of the consumer group dies
+     * <li>A new member is added to an existing consumer group via the join API
+     * </ul>
+     *
+     * @param pattern Pattern to subscribe to
+     */
+    @Override
+    public void subscribe(Pattern pattern) {
+        acquire();
+        try {
+            log.debug("Subscribed to pattern: {}", pattern);
+            this.subscriptions.subscribe(pattern);
+
+            if (this.patternSubscriptionTask == null) {
+                schedulePatternSubscriptionTask(this.patternSubscriptionIntervalMs);
+            }
+        } finally {
+            release();
+        }
+    }
+
+    private void schedulePatternSubscriptionTask(final long interval) {
+        this.patternSubscriptionTask = new DelayedTask() {
+            public void run(long now) {
+                List<String> topicsToSubscribe = new ArrayList<>();
+                for (String topic: listTopics().keySet()) {
```

Comment:
Hey @hachikuji , even I was initially thinking to have it asynchronous, but I could not convince myself that we will gain much with it? schedulePatternSubscriptionTask is already a delayed task that user do not have to wait on. Moreover initiating a new metadata fetch directly needs to handle retries, so more params for max retries, interval, etc. I was under impression that we made ListTopics handle all that so that we will not have to worry about it again during regex implementation. I might be missing something.